### DISCUSSION TITLE
Setting up Pimcore for core contribution

### DISCUSSION THREAD

**Discussion by cngJo** [pimcore/pimcore#17021] [Q&A / Support] [answered]
Maybe it's on me and I cannot use Google or read docs anymore, but I didn't find any way to set up Pimcore for "Core Development" 🤔

I'd like to contribute a small Feature to Pimcore but didn't find any way to get started setting up Pimcore from the `pimcore/pimcore` repository. All supported options require Pimcore as a dependency, which you'd want for production setups but not when contributing to the core itself. 

Unfortunately, the [`CONTRIBUTING.md`](https://github.com/pimcore/pimcore/blob/11.x/CONTRIBUTING.md) wasn't helpful to me in this regard either. 
Also I didn't find any other questions / guides about this topic online. 

Am I missing something obvious? Can anyone help me out here? 
Sure, I could just use `pimcore/skeleton`, do the changes in the vendor directory and then sync them to this repo and commit them, but that feels wrong. 

Thank you!

**Comment by dpfaffenbauer** [comment]
I am using Pimcore for 12 years now and made 100 of contributions and I still don't know what the Pimcore way of dealing with that is.

I believe they just set up the skeleton and install pimcore/pimcore in a lib folder and use composer repositories to link it to that path. 

With that you can then also install all other non core bundles like Admin or Studio.

But that still is a hideous task to setup and also quite error prone. I ended up using pimcore/pimcore and just changing code, that I don't commit, to make it work.

I would love to have a separate documentation and Pimcore's views on how they handle it.

PS: with CoreShop you can just clone the repo, start docker and you are good to go. Makes it super easy to use setup and develop.

**Reply by cngJo** [reply]
Thanks for your reply!

> PS: with CoreShop you can just clone the repo, start docker and you are good to go. Makes it super easy to use setup and develop.

That's more or less on track with what I know from TYPO3 or Shopware, where you basically can clone the "main" repo, "composer install" and start hacking. 

But I also like the approach Pimcore has (https://github.com/orgs/pimcore/discussions/17021#discussioncomment-9346132), since you directly benefit from all the content in `pimcore/demo` which you might not get from other systems, which throw you in the deep end with a blank installation.

**Comment by fashxp** [comment, marked as answer]
Hi, 
we use our [demo](https://github.com/pimcore/demo/) repository. 
Just install the demo as described, then clone your forks of the pimcore repositories you want to contribute to into `dev/pimcore/*` and run `composer update`.
Then, composer should symlink the repositories instead of downloading them into vendor folder and you can make your changes, test them and create PRs. The key part is [here in composer.json](https://github.com/pimcore/demo/blob/11.x/composer.json#L66).

Cheers...

**Reply by cngJo** [reply]
Thank you, that worked for me! 🎉

**Comment by blankse** [comment]
> Hi, we use our [demo](https://github.com/pimcore/demo/) repository. Just install the demo as described, then clone your forks of the pimcore repositories you want to contribute to into `dev/pimcore/*` and run `composer update`. Then, composer should symlink the repositories instead of downloading them into vendor folder and you can make your changes, test them and create PRs. The key part is [here in composer.json](https://github.com/pimcore/demo/blob/11.x/composer.json#L66).
> 
> Cheers...

Interesting approach 👍🏼 

I used following composer.json setting to load the pimcore packages from source:

```
{
  "config": {
    "preferred-install": {
      "pimcore/*": "source"
    }
  }
}
```

Then I can add the fork repositories as remote and make changes directly in vendor/pimcore/*.
